### ISSUE TITLE
Accuracy calculation in Bayesian Neural Network example

### ISSUE THREAD

**Issue by malharjajoo** [tensorflow/probability#711]
In the Bayesian Neural Network example code ([this line](https://github.com/tensorflow/probability/blob/563cdd7bb46e3ce0bd069c713c581e7409f26de1/tensorflow_probability/examples/bayesian_neural_network.py#L269)) provided by TFP,

Why are cheap predictions (cheap = output of **single** forward pass) used to calculate the accuracy of the network during training ? In a BNN, Isn't the mean of multiple forward passes supposed to be used for calculation of accuracy ? 

As a side note, I can see multiple forward passes being used during visualization of outputs for validation set ([this line](https://github.com/tensorflow/probability/blob/563cdd7bb46e3ce0bd069c713c581e7409f26de1/tensorflow_probability/examples/bayesian_neural_network.py#L316))

**Comment by nbro** [comment]
> In a BNN, Isn't the mean of multiple forward passes supposed to be used for calculation of accuracy?

Not necessarily. Each weight of a Bayesian neural network is represented as a distribution. When you perform the forward pass, you sample from this distribution, so the forward pass is stochastic rather than deterministic, in the sense that different forward passes, with the same distributions, can produce different outputs.

In the case of MC dropout, this stochasticity is produced by performing multiple forward passes, but this is a different story.

**Comment by malharjajoo** [comment]
Are you really sure ?  Do you have any sources please ?

**Comment by nbro** [comment]
@malharjajoo I also read [that article](https://nbviewer.jupyter.org/github/krasserm/bayesian-machine-learning/blob/master/bayesian_neural_networks.ipynb). It uses multiple forward passes to have an idea of the variation of the outputs. In fact, you have also noticed this in the TFP example that you mentioned.

> As a side note, I can see multiple forward passes being used during **visualization of outputs for validation set**

> Are you really sure ? Do you have any sources please ?

I've read multiple papers on the topic. I don't even recall having seen the calculation of the accuracy in any of these papers.

**Comment by malharjajoo** [comment]
Ok, my question was indeed about calculation of accuracy during training phase,

 _but I expected the same method_ (single forward pass, but as we see, most examples use multiple forward passes to obtain mean, variance, etc during **test** phase) of calculation for accuracy to be used during both training and testing. 

but I guess I agree with you, since you mentioned MCDropout (and I'm sure it won't use multiple forward passes to calculate accuracy during training)

**Comment by nbro** [comment]
@malharjajoo Anyway, an average accuracy (calculated by performing multiple forward passes) may actually be a more descriptive metric of the performance of the model. AFAIK, there hasn't been much work on the creation of metrics for assessing the performance of BNNs.
